Haptics/app/src/main/java/com/example/android/haptics/samples/ui/resist/ResistRoute.kt

```diff
@@ -0,0 +1,335 @@
+
+    // Actual values used to draw indicator (since we want to animate).
+    // Note we only need an animation back, otherwise animation moves as the user drags.
+    val animationSpecDp: AnimationSpec<Dp> = if (isDragging) tween(0) else tween(TIME_TO_ANIMATE_BACK_MS)
+    val animationSpecFloat: AnimationSpec<Float> = if (isDragging) tween(0) else tween(TIME_TO_ANIMATE_BACK_MS)
+    val sizeAnimated by animateDpAsState(
+        targetValue = if (isDragging) size else START_SIZE,
+        animationSpec = animationSpecDp
+    )
+    val strokeWidthAnimated by animateDpAsState(
+        targetValue = if (isDragging) strokeWidth else START_STROKE_WIDTH,
+        animationSpec = animationSpecDp
+    )
+    val rotationAnimated by animateFloatAsState(
+        targetValue = if (isDragging) rotation else START_INITIAL_ROTATION,
+        animationSpec = animationSpecFloat
+    )
+    val offsetYAnimated by animateDpAsState(
+        targetValue = if (isDragging) offsetY else START_Y_OFFSET,
+        animationSpec = animationSpecDp
+    )
+
+    // The interval and the intensity of the vibrations.
+    var vibrationInterval by remember { mutableStateOf(TICK_INTERVAL_MAX_MS) }
+    var vibrationIntensity by remember { mutableStateOf(TICK_INTENSITY_MIN) }
+
+    // Use an interpolator to simulate resistance as the user approaches end of their drag.
+    val interpolator = DecelerateInterpolator()
+
+    // Use the drag distance to calculate the indicator state and the vibration intensity.
+    if (dragOffset <= 0) {
+        // Initial state of the indicator.
+        offsetY = START_Y_OFFSET
+        size = START_SIZE
+        strokeWidth = START_STROKE_WIDTH
+        rotation = START_INITIAL_ROTATION
+
+        vibrationInterval = TICK_INTERVAL_MAX_MS
+        vibrationIntensity = TICK_INTENSITY_MIN
+    } else if (dragOffset < maxDragOffset) {
+        // User is dragging between start and max drag distance.
+        val percentDragInterpolated = interpolator.getInterpolation(dragOffset / maxDragOffset)
+        val relativeInterval =
+            ((1f - percentDragInterpolated) * (TICK_INTERVAL_MAX_MS - TICK_INTERVAL_MIN_MS)).toLong()
+        val relativeIntensity =
+            percentDragInterpolated * (TICK_INTENSITY_MAX - TICK_INTENSITY_MIN)
+        val additionalSizeBasedOnDrag = Dp(percentDragInterpolated * (TARGET_SIZE.value - START_SIZE.value))
+        val additionalStrokeWidthBasedOnDrag =
+            Dp((percentDragInterpolated * (TARGET_STROKE_SIZE.value - START_STROKE_WIDTH.value)))
+        val additionalRotation = percentDragInterpolated * TARGET_ROTATION
+        offsetY = Dp(percentDragInterpolated * TARGET_Y_OFFSET.value)
+        size = START_SIZE + additionalSizeBasedOnDrag
+        strokeWidth = START_STROKE_WIDTH + additionalStrokeWidthBasedOnDrag
+        rotation = additionalRotation + START_INITIAL_ROTATION
+
+        vibrationInterval = TICK_INTERVAL_MIN_MS + relativeInterval
```

Comment:
The code for (start + relative * (end - start)) is quite repeated here.

Please consider using MathUtils.lerp to linear interpolate between min/max values using a single relativeDragOffset variable, you can then cast/convert to each required unit the result.

Reply:
Done, didn't seem like there was a function available to me, so added a MathUtils file in utils package